amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,422 @@
+#[derive(Debug, Clone)]
+pub struct Causes<'a> {
+    current: Option<&'a Error>,
+}
+
+impl<'a> Iterator for Causes<'a> {
+    type Item = &'a Error;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(e) = self.current {
+            self.current = e.source();
+            return Some(e);
+        }
+
+        None
+    }
+}
+
+/// Constructs an `Error` using the standard string interpolation syntax.
+///
+/// ```rust
+/// #[macro_use] extern crate amethyst_error;
+///
+/// fn main() {
+///     let err = format_err!("number: {}", 42);
+///     assert_eq!("number: 42", err.to_string());
+/// }
+/// ```
+#[macro_export]
+macro_rules! format_err {
+    ($($arg:tt)*) => { $crate::Error::from_string(format!($($arg)*)) }
+}
+
+/// Constructs an [`Error`](Error) from a string.
+pub fn err_msg<M>(message: M) -> Error
+where
+    M: 'static + Send + Sync + fmt::Debug + fmt::Display,
+{
+    Error::new(ErrMsgError { message })
+}
+
+/// Treat something that can be displayed as an error.
+struct ErrMsgError<M> {
+    message: M,
+}
+
+impl<M> error::Error for ErrMsgError<M> where M: 'static + Send + Sync + fmt::Debug + fmt::Display {}
+
+impl<M> fmt::Display for ErrMsgError<M>
+where
+    M: fmt::Display,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.message.fmt(fmt)
+    }
+}
+
+impl<M> fmt::Debug for ErrMsgError<M>
+where
+    M: fmt::Debug,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.message.fmt(fmt)
+    }
+}
+
+#[cfg(not(all(feature = "backtrace", feature = "std")))]
+mod internal {
+    use std::fmt;
+
+    /// Fake internal representation.
+    pub struct Backtrace(());
+
+    impl fmt::Debug for Backtrace {
+        fn fmt(&self, _fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+            Ok(())
+        }
+    }
+}
+
+#[cfg(all(feature = "backtrace", feature = "std"))]
+mod internal {
+    pub use backtrace::Backtrace;
+    use std::{env, ffi, sync::atomic};
+
+    const RUST_BACKTRACE: &str = "RUST_BACKTRACE";
+
+    /// Test if backtracing is enabled.
+    pub fn is_backtrace_enabled<F: Fn(&str) -> Option<ffi::OsString>>(get_var: F) -> bool {
+        match get_var(RUST_BACKTRACE) {
+            Some(ref val) if val != "0" => true,
+            _ => false,
+        }
+    }
+
+    /// Constructs a new backtrace, if backtraces are enabled.
+    fn new() -> Option<Backtrace> {
+        static ENABLED: atomic::AtomicUsize = atomic::ATOMIC_USIZE_INIT;
+
+        match ENABLED.load(atomic::Ordering::Relaxed) {
+            0 => {
+                let enabled = is_backtrace_enabled(|var| env::var_os(var));
+
+                ENABLED.store(enabled as usize + 1, atomic::Ordering::Relaxed);
+
+                if !enabled {
+                    return None;
+                }
+            }
+            1 => return None,
+            _ => {}
+        }
+
+        Some(Backtrace::new())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
We should add test to check backtrace functionality.

Reply:
Done!